docs/design/branch-mode.md

```diff
@@ -0,0 +1,117 @@
+# Branch mode
+
+This is a proposed configuration option to replace
+`experimental-advance-branches`. It is not intended to replace or obviate
+topics.
```

Comment:
I think it would help me if this document started with a problem statement. Can you list some scenarios that you think we should improve? Here's one I can think of where it should be easy to make jj behave better:

```sh
$ jj
@  n martinvonz@google.com 2024-10-19 17:22:54 70e302a0
│  (no description set)
○  w martinvonz@google.com 2024-10-19 17:05:55 main HEAD@git 873ebde8
│  (empty) (no description set)
○  u martinvonz@google.com 2024-10-19 17:05:47 52b0bdfd
│  (empty) (no description set)
◆  z root() 00000000

$ git branch
* main

$ jj desc w -m blah
Rebased 1 descendant commits
Working copy now at: n e270b828 (no description set)
Parent commit      : w a316f16d main | (empty) blah

$ git branch
* (HEAD detached at a316f16)
  main
```

I.e. we left Git in a "detached HEAD" state where we could have left `main` checked out.

Reply:
Since `describe` doesn't move `@` (it changes the commit ID but not the change ID), it also wouldn't change `.git/HEAD` in "active branch" mode, since branch mode means that `.git/HEAD` contains a branch reference, not a commit sha.

(As before, I will add this to the doc, I just wanted to briefly answer your question now.)

Reply:
> Since `describe` doesn't move `@` (it changes the commit ID but not the change ID), it also wouldn't change `.git/HEAD` in "active branch" mode, since branch mode means that `.git/HEAD` contains a branch reference, not a commit sha.
> 
> (As before, I will add this to the doc, I just wanted to briefly answer your question now.)

But the Git branch that HEAD@git points to always points to a commit hash, not a change_id, so it does need updating. Or, in terms of bookmarks, the JJ bookmark might still point to the same change_id, but as a newly  created commit is associated with the change_id from the bookmark, and the old commit becomes hidden, JJ effectively needs to run `jj git export` to update the commit hash that git sees for the branch that the bookmark is tracking.

Reply:
The proposal is currently only for colocated repos. In colocated repos, the export already happens automatically:

> Jujutsu will import and export from and to the Git repo on every jj command automatically.

I'll need to learn a little more about how non-colocated repos currently behave before adding them to the design.